src/test/main.cpp

```diff
@@ -0,0 +1,1154 @@
+
+// SimpleIni library is licensed to MIT
+// xxHash library is licensed to BSD-2-Clause
+
+#include <clocale>
+#include <cstdint>
+#include <cstdlib>
+#include <filesystem>
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <iterator>
+#include <map>
+#include <sstream>
+#include <stdlib.h>
+#include <string>
+// TODO: Need fix multi-threading
+// Known bug from XbSymbolDatabase:
+// * Console log appear to only output one at a time from one thread to next.
+// * Sometimes get error'd for certain libraries. Some sort of conflict occurred.
+//#define ENABLE_MULTI_THREAD
+#ifdef ENABLE_MULTI_THREAD
+#include <thread>
+#endif
+
+#include <libXbSymbolDatabase.h>
+#include <xref/list_xref.h>
+#include <SimpleIni.h>
+#include <xxhash.h>
+
+bool operator==(const CSimpleIni::Entry& i, const CSimpleIni::Entry& ii)
+{
+    return std::strcmp(i.pItem, ii.pItem) == 0;
+}
+
+#include "Xbe.h"
+#include "util/helper.hpp"
+#include "libverify.hpp"
+#include "util/cliConfig.hpp"
+
+#define _128_MiB 0x08000000
+
+std::map<uint32_t, symbol_result> g_SymbolAddresses;
+std::ios_base::fmtflags cout_fmt = std::cout.flags();
+unsigned int XbSDB_test_error = 0;
+unsigned int XbUnitTest_error = 0;
+
+static const char* cli_argument_str = "> XbSymbolUnitTest"
+                                      " [-h|--help]"
+                                      " default.xbe"
+                                      " [-out|--out <output to specific folder>]"
+                                      " [-f]"
+                                      " [-v|--verbose]\n";
+
+#define UNITTEST_OK 0
+#define UNITTEST_FAIL_INVALID_ARG 1
+#define UNITTEST_FAIL_XBSDB 2
+#define UNITTEST_FAIL_OPEN_FILE 3
+#define UNITTEST_FAIL_INVALID_XBE 4
+#define UNITTEST_FAIL_UNABLE_ALLOC_MEM 5
+#define UNITTEST_FAIL_SYMBOLS_NOT_FOUND 6
+#define UNITTEST_FAIL_SYMBOLS_DIFF_SIZE 7
+#define UNITTEST_FAIL_SYMBOLS_NOT_MATCH 8
+#define UNITTEST_FAIL_DATABASE_NOT_SYNC 9
+
+// Buffer generation result in case user want to save as cache file.
+// Which will be use to compare against generated cache result.
+CSimpleIni gen_result(false, true, false);
+
+static const char* section_XbSDb_messages = "XbSDb Messages";
+static const char* section_XbSUT_messages = "XbSUT Messages";
+static struct {
+    const char* unknown = "UNKNOWN";
+    const char* warn = "WARN";
+    const char* error = "ERROR";
+} sect_generic_messages;
+
+static const char* section_certificate = "Certificate";
+static struct {
+    const char* name = "Name";
+    const char* TitleID = "TitleID";
+    const char* TitleIDHex = "TitleIDHex";
+    const char* Region = "Region";
+} sect_certificate;
+
+static const char* section_libs = "Libs";
+static struct {
+    const char* BuildVersion = "BuildVersion";
+    const char* DSOUND = "DSOUND";
+    const char* JVS = "JVS";
+} sect_libs;
+
+static void pause_for_user_input(void)
+{
+    std::cout << "Press 'enter' key to continue...";
+    (void)std::getchar();
```

Comment:
I'm not sure why you'd want this. It prevents easy automatization (for example, on CI, where no user is present) and doesn't add much benefit.

Reply:
On CI, there's a limit of 10 seconds for the ctest to complete or become considered as failure. You can find this in c42e1d599e0caee1f676e735964e61362a68621c commit.

It actually add benefit for multiple titles testing purpose. From my own drag 'n drop to cache file, which contain 80+ title xbes, only 6 terminals remain opened because there are errors with the signatures occur. Also, if there's a problem with certain titles which may not stored in cache file. They will be easily missed during testings.

Reply:
But shouldn't such errors be put in a separate log (or at least `stderr`) to prevent user-interaction?

Reply:
Separate log for errors? Isn't that unwise for 900+ titles? For drivers, daily software, etc, sure. Except original xbox titles are way too many to keep in a separate log. Using cache folder can keep warnings and errors report cached and can be compare against new test's changes. While Windows' command prompt is still opened, I can perform a search for any error found which can be seen at the end of the test for error counts.

Reply:
I had struggled to remember what was common cli used for non-interactive mode until recently. Using `-i` or `--interactive` can solve this problem. I will start write up implement for it. Any non-critical features request should be done in separate pull request or in a ticket.